fiat-roles/src/main/java/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepository.java

```diff
@@ -76,43 +82,120 @@
   private static final String UNRESTRICTED = UnrestrictedResourceConfig.UNRESTRICTED_USERNAME;
   private static final String NO_LAST_MODIFIED = "unknown_last_modified";
 
+  private final Clock clock;
   private final ObjectMapper objectMapper;
   private final RedisClientDelegate redisClientDelegate;
   private final List<Resource> resources;
+  private final RedisPermissionRepositoryConfigProps configProps;
+  private final RetryRegistry retryRegistry;
+
   private final LoadingCache<String, UserPermission> unrestrictedPermission =
       Caffeine.newBuilder()
           .expireAfterAccess(Duration.ofSeconds(10))
           .build(k -> reloadUnrestricted());
 
   private final String prefix;
 
-  public RedisPermissionsRepository(
+  RedisPermissionsRepository(
+      Clock clock,
       ObjectMapper objectMapper,
       RedisClientDelegate redisClientDelegate,
       List<Resource> resources,
-      String prefix) {
+      RedisPermissionRepositoryConfigProps configProps,
+      RetryRegistry retryRegistry) {
+    this.clock = clock;
     this.objectMapper = objectMapper;
     this.redisClientDelegate = redisClientDelegate;
-    this.prefix = prefix;
+    this.configProps = configProps;
+    this.prefix = configProps.getPrefix();
     this.resources = resources;
+    this.retryRegistry = retryRegistry;
+  }
+
+  public RedisPermissionsRepository(
+      ObjectMapper objectMapper,
+      RedisClientDelegate redisClientDelegate,
+      List<Resource> resources,
+      RedisPermissionRepositoryConfigProps configProps,
+      RetryRegistry retryRegistry) {
+    this(
+        Clock.systemUTC(),
+        objectMapper,
+        redisClientDelegate,
+        resources,
+        configProps,
+        retryRegistry);
   }
 
   private UserPermission reloadUnrestricted() {
-    return getFromRedis(UNRESTRICTED).orElseThrow(NoSuchElementException::new);
+    return getFromRedis(UNRESTRICTED)
+        .orElseThrow(() -> new PermissionRepositoryException("Failed to read unrestricted user"));
   }
 
   private UserPermission getUnrestrictedUserPermission() {
     String serverLastModified =
-        redisClientDelegate.withCommandsClient(
-            c -> {
-              return c.get(unrestrictedLastModifiedKey());
-            });
+        redisRead(
+            new TimeoutContext(
+                "checkLastModified",
+                clock,
+                configProps.getRepository().getCheckLastModifiedTimeout()),
+            c -> c.get(unrestrictedLastModifiedKey()));
     if (serverLastModified == null) {
       serverLastModified = NO_LAST_MODIFIED;
     }
     return unrestrictedPermission.get(serverLastModified);
   }
 
+  private static class TimeoutContext {
```

Comment:
I'll add a doc comment on this class, but essentially it is here to bail on trying to load a user if too much elapsed time has passed. Because this is doing retries at the individual redis read level instead of a retry around the whole get method in the case of connection pool contention or other redis slowness I don't want to eventually succeed if the result is useless at that point - it would be better to bail out with an error and let external retries or fallbacks kick in